elki-outlier/src/main/java/elki/outlier/density/HySortOD.java

```diff
@@ -0,0 +1,743 @@
+ * for fast outlier detection.
+ * Its main focus is the analysis of data with many instances and a
+ * low-to-moderate number of dimensions.
+ * <p>
+ * Reference:
+ * <p>
+ * Cabral, Eugênio F., Robson L.F. Cordeiro<br>
+ * Fast and Scalable Outlier Detection with Sorted Hypercubes<br>
+ * Proc. 29th ACM Int. Conf. on Information & Knowledge Management (CIKM'20)
+ * 
+ * @author Cabral, Eugênio F. (Original Code)
+ * @author Braulio V.S. Vinces (ELKIfication)
+ * @since 0.7.5
+ *
+ * @param <V> Vector type
+ */
+@Title("HySortOD: Hypercube-Based Outlier Detection")
+@Description("Algorithm that uses an efficient hypercube-ordering-and-searching strategy for fast outlier detection.")
+@Reference(authors = "Cabral, Eugênio F., Robson L.F. Cordeiro", //
+    title = "Fast and Scalable Outlier Detection with Sorted Hypercubes", //
+    booktitle = "Proc. 29th ACM Int. Conf. on Information & Knowledge Management (CIKM'20)", //
+    url = "https://doi.org/10.1145/3340531.3412033")
+@Alias({ "de.lmu.ifi.dbs.elki.algorithm.outlier.HySortOD", "hysort" })
+public class HySortOD<V extends NumberVector> implements OutlierAlgorithm {
+
+  /**
+   * The logger for this class.
+   */
+  private static final Logging LOG = Logging.getLogger(HySortOD.class);
+
+  /**
+   * Number of bins
+   */
+  protected int b;
+
+  /**
+   * Threshold to balance the trade-off between the number of hyper-cubes to
+   * scan
+   * during the search and the granularity of mapping
+   */
+  protected int minSplit;
+
+  // hypercube's length
+  private final double l;
+
+  // to match each object with its hyper-cube
+  // it's a little trick to deal with hypercube rearrangement
+  private int[] invertedIndex;
+
+  // strategy for hyper-cubes density search
+  private DensityStrategy strategy;
+
+  /**
+   * Constructor
+   * 
+   * @param k
+   * @param b Number of bins
+   */
+  public HySortOD(int b, int minSplit) {
+    super();
+    this.b = b;
+    this.l = 1 / (double) this.b;
+    this.minSplit = minSplit;
+    if(this.minSplit > 0) 
+      this.strategy = new TreeStrategy(this.minSplit);
+     else 
+      this.strategy = new NaiveStrategy();
+  }
+
+  public OutlierResult run(Database db, Relation<V> relation) {
+
+    // Output data storage
+    WritableDoubleDataStore scores = DataStoreUtil.makeDoubleStorage(relation.getDBIDs(), DataStoreFactory.HINT_DB);
+    FiniteProgress prog = LOG.isVerbose() ? new FiniteProgress("HySortOD scores", relation.size(), LOG) : null;
+    // Track minimum and maximum scores
+    DoubleMinMax minmax = new DoubleMinMax();
+
+    int[] W;
+    List<Hypercube> H;
+
+    {
+      H = getSortedHypercubes(relation);
+      W = this.strategy.buildIndex(H).getDensities();
+
+      // Iterate over all objects
+      int objectPosition = 0;
+      for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance())
+        scores.putDouble(iter, score(W[this.invertedIndex[objectPosition++]]));
+
+      LOG.incrementProcessed(prog);
```

Comment:
Clearly, this progress logging was meant to be in the loop, or it will be executed only once...

Reply:
But I guess most work is done in the hypercube methods above, a progress bar would be more helpful there.